### PR TITLE
[TreeView] Make each plugin responsible for its context value

### PR DESCRIPTION
**Pull request by flaviendelangle**
Extracted from #11452

The idea here is to have each plugin decide what to pass to the context and to type it.

### CODE REVIEW

File: packages/x-tree-view/src/internals/models/helpers.ts [hunk 1]
```diff
@@ -8,6 +8,14 @@ export type DefaultizedProps<
   Required<Pick<P, RequiredProps>> &
   AdditionalProps;
 
+type IsAny<T> = 0 extends 1 & T ? true : false;
```
Comment (michelengelen) [inline packages/x-tree-view/src/internals/models/helpers.ts:11]: NICE

**Review by michelengelen — APPROVED** [review]
NICE work!!! 🚀

File: packages/x-tree-view/src/internals/models/helpers.ts (hunk 1 above)
Comment (flaviendelangle) [reply packages/x-tree-view/src/internals/models/helpers.ts:11]: Found that one on Stackoverflow :laughing:

Comment (michelengelen) [reply packages/x-tree-view/src/internals/models/helpers.ts:11]: I had to read an explanation article about it to understand that! :D

**Pull request merged**